ui-trackoccupancydiagram/src/components/helpers/drawElements/drawOccupancyZones.ts

```diff
@@ -4,58 +4,197 @@ import {
+  ctx.moveTo(arrivalTime - ARROW_OFFSET_X, OCCUPANCY_ZONE_START + ARROW_OFFSET_Y);
+  ctx.lineTo(arrivalTime - ARROW_WIDTH, OCCUPANCY_ZONE_START - ARROW_TOP_Y);
+  ctx.lineTo(arrivalTime + ARROW_WIDTH, OCCUPANCY_ZONE_START - ARROW_TOP_Y);
+  ctx.lineTo(arrivalTime + ARROW_OFFSET_X, OCCUPANCY_ZONE_START + ARROW_OFFSET_Y);
+  ctx.lineTo(arrivalTime + ARROW_WIDTH, OCCUPANCY_ZONE_START + ARROW_BOTTOM_Y);
+  ctx.lineTo(arrivalTime - ARROW_WIDTH, OCCUPANCY_ZONE_START + ARROW_BOTTOM_Y);
+  ctx.lineTo(arrivalTime - ARROW_OFFSET_X, OCCUPANCY_ZONE_START + ARROW_OFFSET_Y);
+  ctx.fill();
+  ctx.moveTo(arrivalTime - ARROW_OFFSET_X, OCCUPANCY_ZONE_START + ARROW_OFFSET_Y);
+  ctx.lineTo(arrivalTime + ARROW_OFFSET_X, OCCUPANCY_ZONE_START + ARROW_OFFSET_Y);
+  ctx.stroke();
+};
+
+type DrawRemainingTrainsBox = {
+  ctx: CanvasRenderingContext2D;
+  text: string;
+  textX: number;
+  textY: number;
+};
+
+const drawRemainingTrainsBox = ({ ctx, text, textX, textY }: DrawRemainingTrainsBox) => {
+  ctx.fillStyle = REMAINING_TRAINS_BACKGROUND;
+  ctx.beginPath();
+  ctx.rect(textX, textY, REMAINING_TRAINS_WIDTH, REMAINING_TRAINS_HEIGHT);
+  ctx.fill();
+  ctx.stroke();
+  ctx.fillStyle = WHITE_100;
+  ctx.font = SANS;
+  ctx.textAlign = 'center';
+  ctx.textBaseline = 'middle';
+  ctx.fillText(text, textX + REMAINING_TRAINS_WIDTH / 2, textY + REMAINING_TRAINS_HEIGHT / 2);
+};
+
 export const drawOccupancyZones = ({
   ctx,
   width,
   height,
   tracks,
   occupancyZones,
   getTimePixel,
+  selectedTrainId,
 }: {
   ctx: CanvasRenderingContext2D;
   width: number;
   height: number;
   tracks: Track[] | undefined;
   occupancyZones: OccupancyZone[] | undefined;
   getTimePixel: (time: number) => number;
+  selectedTrainId: string;
 }) => {
   ctx.clearRect(0, 0, width, height);
 
-  tracks?.forEach((track, index) => {
+  if (!tracks || !occupancyZones || occupancyZones.length === 0) return;
+
+  const sortedOccupancyZone = occupancyZones.sort(
+    (a, b) => a.arrivalTime.getTime() - b.arrivalTime.getTime()
+  );
+
+  tracks.forEach((track, index) => {
     const trackTranslate = index === 0 ? CANVAS_PADDING : TRACK_HEIGHT_CONTAINER;
     ctx.translate(0, trackTranslate);
 
-    const trackOccupancyZones = occupancyZones?.filter((zone) => zone.trackId === track.id);
+    const filteredOccupancyZone = sortedOccupancyZone.filter((zone) => zone.trackId === track.id);
 
-    if (!trackOccupancyZones) return;
+    let primaryArrivalTimePixel = getTimePixel(filteredOccupancyZone[0].arrivalTime.getTime());
+    let primaryDepartureTimePixel = getTimePixel(filteredOccupancyZone[0].departureTime.getTime());
+    let lastDepartureTimePixel = primaryDepartureTimePixel;
+    let yPosition = OCCUPANCY_ZONE_START;
+    let yOffset = 0;
+    let counter = 0;
 
-    trackOccupancyZones.forEach((zone) => {
+    // use a for of loop to be able to break the loop if there are more than 9 zones
+    for (const [sortedIndex, zone] of filteredOccupancyZone.entries()) {
```

Comment:
Can you rename `index` l104 => `trackIndex` and `sortIndex` => `index` ? Or smth like this. `sortIndex` is a weird name
```suggestion
    for (const [index, zone] of filteredOccupancyZone.entries()) {
```

Reply:
replaced by `zoneIndex`